snapcraft/internal/pluginhandler/__init__.py

```diff
@@ -227,14 +222,21 @@ def mark_cleaned(self, step):
             os.rmdir(self.statedir)
 
     def _fetch_stage_packages(self):
-        try:
-            self.stage_packages = self._stage_package_handler.fetch()
-        except repo.errors.PackageNotFoundError as e:
-            raise RuntimeError("Error downloading stage packages for part "
-                               "{!r}: {}".format(self.name, e.message))
+        stage_packages = self._grammar_processor.get_stage_packages()
+        if stage_packages:
+            logger.debug('Fetching stage-packages {!r}'.format(stage_packages))
+            try:
+                self.stage_packages = self._stage_packages_repo.get(
+                    stage_packages)
+            except repo.errors.PackageNotFoundError as e:
+                raise errors.StagePackageDownloadError(self.name, e.message)
 
     def _unpack_stage_packages(self):
-        self._stage_package_handler.unpack(self.installdir)
+        stage_packages = self._grammar_processor.get_stage_packages()
```

Comment:
you can process these only once with a `@property` trick or using `get_stage_packages`, in both cases storing the result

Reply:
nvm, this might not be necessary at all